docs/developer-docs/security/rust-canister-development-security-best-practices.mdx

```diff
@@ -831,6 +831,145 @@ Publicly revealing the canister's cycles balance allows an attacker to measure t
+async fn seed_randomness() {
+    let (seed,): ([u8; 32],) = ic_cdk::call(Principal::management_canister(), "raw_rand", ())
+        .await
+        .expect("Failed to call the management canister");
+    RNG.with_borrow_mut(|rng| *rng = Some(ChaCha20Rng::from_seed(seed)));
+}
+
+fn schedule_seeding(duration: Duration) {
+    ic_cdk_timers::set_timer(duration, move || {
+        ic_cdk::spawn(async move {
+            seed_randomness().await;
+            schedule_seeding(duration);
+        })
+    });
+}
+
+#[ic_cdk::init]
+fn init() {
+    // Initialize randomness during canister install or reinstall
+    ic_cdk_timers::set_timer(Duration::ZERO, || {
+        ic_cdk::spawn(async {
+            seed_randomness().await;
+            // Schedule reseeding on a timer with duration SEEDING_INTERVAL
+            schedule_seeding(SEEDING_INTERVAL);
+        })
+    });
+}
+
+#[ic_cdk::post_upgrade]
+fn post_upgrade() {
+    // Initialize randomness after a canister upgrade
+    ic_cdk_timers::set_timer(Duration::ZERO, || {
+        ic_cdk::spawn(async {
+            seed_randomness().await;
+            // Schedule reseeding on a timer with duration SEEDING_INTERVAL
+            schedule_seeding(SEEDING_INTERVAL);
+        })
+    });
+}
+
+// This must always be an update method or the PRNG state won't be updated
+#[ic_cdk::update]
+fn random_bytes(size: u32) -> RandomBytesResult {
+    let mut buf = vec![0; size as usize];
+    RNG.with_borrow_mut(|rng| match rng.as_mut() {
+        Some(rand) => {
+            rand.fill_bytes(&mut buf);
+            Ok(hex::encode(buf))
+        }
+        None => Err(RngError::RngNotInitialized(
+            "Randomness is not initialized. Please try again later".to_string(),
+        )),
+    })
+}
+```
+
+##### Benefits:
+- This method scales for large random bytes as `raw_rand` needs be called only once and subsequent PRNG computation is local to the canister.
+
+##### Caveats:
+- The `setup_randomness` must **always** be intialized in both the `init` and `post_upgrade` hook as `init` is not invoked during an canister upgrade [see here](https://internetcomputer.org/docs/current/references/ic-interface-spec#system-api-upgrades).
+- The `init` and `post_upgrade` methods doesn't allow async calls and thus a timer is immediately scheduled to seed the randomness.
+- Once the seed is intialized, the outcome of all future `random_bytes` is predicable to anyone having the seed (node providers) as the PRNG is deterministic. This breaks the unpredicatable property of secure randomness. Hence, to balance security vs performance, we recommend frequently reseeding the PRNG on a timer. The example above already does this with duration of **1 hour**. However, based on the sensitivity of their dApp, developers can choose an appropriate reseeding interval by setting `SEEDING_INTERVAL`.
+- The `random_bytes` must **always** be an `update` method, so the PRNG can preserve the state and offer unique randomness on every request.
```

Comment:
I guess the difference is the motoko method can be never called as a query as the canister would trap. Here, it wouldn't.

Reply:
Nope, if people use `#[ic_cdk::query]` instead of `#[ic_cdk::update]`, everything would work the same but the `random_bytes` would be the same on every call. Hence, I felt it's important to highlight it.